{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.01\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3\n",
    "#\n",
    "# Moritz Blumer, 2024\n",
    "\n",
    "\"\"\"\n",
    "WinPCA. A package for windowed PC analysis.\n",
    "\"\"\"\n",
    "\n",
    "## IMPORT MODULES\n",
    "from modules import config\n",
    "from modules.cli import CLI\n",
    "# from modules.error_handling import ErrorHandling ## add"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Namespace(flip_pc='1', flip_windows='flip_windows.tsv', prefix='test', reflect=True, winpca='flip')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'winpca': 'flip',\n",
       " 'prefix': 'test',\n",
       " 'flip_windows': 'flip_windows.tsv',\n",
       " 'reflect': True,\n",
       " 'flip_pc': '1',\n",
       " 'flip_window_lst': ['500000', '1000000-2000000'],\n",
       " 'skip_monomorphic': False,\n",
       " 'min_var_per_w': 25,\n",
       " 'n_prev_windows': 5,\n",
       " 'pol_pc': 'both',\n",
       " 'chrom_plot_w': 1200,\n",
       " 'chrom_plot_h': 400,\n",
       " 'genome_plot_h': 300}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## DELETE CELL\n",
    "import sys\n",
    "\n",
    "# simulate command line arguments\n",
    "# command_line = 'winpca.ipynb pca -n test --vcf test_dataset/input/sample.vcf.gz -s ind_1,ind_2,ind_3,ind_4,ind_5,ind_6 -r chr1:1-33000000 -w 1000000 -i 100000 -m 0.01 -p auto'\n",
    "command_line = 'winpca.ipynb pca -n test --vcf test_dataset/input/sample.vcf.gz -s samples.tsv -p auto -r chr1:1-30000000 -w 1000000 -i 100000 -m 0.01'\n",
    "# command_line = 'winpca.ipynb polarize -n test -p auto -c 2'\n",
    "#command_line = 'winpca.ipynb chromplot -n test -r chr1:1-30000000 -m test_dataset/input/metadata.tsv -i 10 -f PDF,html -g inversion_state -c ancestral:eb4034,inverted:2f35a8,heterozygous:197d34'\n",
    "command_line = 'winpca.ipynb genomeplot -p genome_plot_test/ -r chr1,chr2,chr3,chr4,chr5 -m test_dataset/input/metadata.tsv -i 10 -f PDF,html -g inversion_state -c ancestral:eb4034,inverted:2f35a8,heterozygous:197d34'\n",
    "# command_line = 'winpca.ipynb pca -h'\n",
    "command_line = 'winpca.ipynb flip -n test --r -w flip_windows.tsv'\n",
    "sys.argv = command_line.split(' ')\n",
    "\n",
    "\n",
    "## PARSE COMMAND LINE ARGUMENTS\n",
    "\n",
    "# instantiate\n",
    "cli = CLI()\n",
    "\n",
    "# call subparsers\n",
    "cli.pca()\n",
    "cli.pcangsd()\n",
    "cli.polarize()\n",
    "cli.flip()\n",
    "cli.chromplot()\n",
    "cli.genomeplot()\n",
    "\n",
    "# parse argments\n",
    "cli.parse_args()\n",
    "args_dct = cli.args_dct\n",
    "\n",
    "# enter mode\n",
    "mode = args_dct['winpca']\n",
    "args_dct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "[INFO] Reading data from prefix \"test*\".\n"
     ]
    }
   ],
   "source": [
    "# WINDOWED PCA FROM CALLED GENOTYPES\n",
    "\n",
    "if mode == 'pca':\n",
    "\n",
    "    # import relevant modules\n",
    "    from modules.windowed_pca import GT_WPCA\n",
    "    from modules.data import WPCA_Data\n",
    "\n",
    "    # instantiate windowed PCA\n",
    "    w_pca = GT_WPCA(\n",
    "        variant_file_path = args_dct['variant_file_path'],\n",
    "        sample_lst = args_dct['sample_lst'],\n",
    "        chrom = args_dct['chrom'],\n",
    "        start = args_dct['start'],\n",
    "        stop = args_dct['end'],\n",
    "        w_size = args_dct['w_size'],\n",
    "        w_step = args_dct['w_step'],\n",
    "        skip_monomorphic=config.SKIP_MONOMORPHIC,\n",
    "        )\n",
    "    \n",
    "    # run\n",
    "    w_pca.win_vcf_gt()\n",
    "\n",
    "    # parse run data\n",
    "    data = WPCA_Data(args_dct['prefix'], w_pca)\n",
    "\n",
    "\n",
    "# WINDOWED PCA FROM CALLED GENOTYPES\n",
    "\n",
    "# elif mode == 'pcangsd':\n",
    "\n",
    "#     [...]\n",
    "\n",
    "# EXISTING DATA:\n",
    "\n",
    "else:\n",
    "    from modules.data import WPCA_Data\n",
    "    data = WPCA_Data(args_dct['prefix'])\n",
    "\n",
    "\n",
    "\n",
    "# POLARIZE\n",
    "\n",
    "if mode in ['pca', 'pcangsd', 'polarize'] \\\n",
    "    and not args_dct['polarize'] == 'skip':\n",
    "    \n",
    "    from modules.transform_data import Polarize\n",
    "    polarize = Polarize()\n",
    "\n",
    "    # adaptive\n",
    "    if args_dct['polarize'] == 'auto':\n",
    "        if args_dct['pol_pc'] == 'both':\n",
    "            data.modify_data(\n",
    "                'pc_1', polarize.adaptive, args_dct['n_prev_windows']\n",
    "            )\n",
    "            data.modify_data(\n",
    "                'pc_2', polarize.adaptive, args_dct['n_prev_windows']\n",
    "            )\n",
    "        else:\n",
    "            data.modify_data(\n",
    "                'pc_' + str(args_dct['pol_pc']), polarize.adaptive, \\\n",
    "                    args_dct['n_prev_windows']\n",
    "            )\n",
    "\n",
    "    # using guide samples\n",
    "    if args_dct['polarize'] == 'guide_samples':\n",
    "        if args_dct['pol_pc'] == 'both':\n",
    "            data.modify_data(\n",
    "                'pc_1', polarize.guide_samples, args_dct['guide_sample_lst']\n",
    "            )\n",
    "            data.modify_data(\n",
    "                'pc_2', polarize.guide_samples, args_dct['guide_sample_lst']\n",
    "            )\n",
    "        else:\n",
    "            data.modify_data(\n",
    "                'pc_' + str(args_dct['pol_pc']), polarize.guide_samples, \\\n",
    "                    args_dct['guide_sample_lst']\n",
    "            )\n",
    "\n",
    "\n",
    "# FLIP\n",
    "\n",
    "if mode == 'flip':\n",
    "    from modules.transform_data import Flip\n",
    "    flip = Flip()\n",
    "    # reflect\n",
    "    if args_dct['reflect']:\n",
    "        if args_dct['flip_pc'] == 'both':\n",
    "            data.modify_data(\n",
    "                'pc_1', flip.flip_chrom\n",
    "            )\n",
    "            data.modify_data(\n",
    "                'pc_2', flip.flip_chrom\n",
    "            )\n",
    "        else:\n",
    "            data.modify_data(\n",
    "                'pc_' + str(args_dct['flip_pc']), flip.flip_chrom\n",
    "            )\n",
    "    # flip specified windows\n",
    "    if args_dct['flip_windows']:\n",
    "        if args_dct['flip_pc'] == 'both':\n",
    "            data.modify_data(\n",
    "                'pc_1', flip.flip_windows, \n",
    "                args_dct['flip_window_lst']\n",
    "            )\n",
    "            data.modify_data(\n",
    "                'pc_2', flip.flip_windows, \n",
    "                args_dct['flip_window_lst']\n",
    "            )\n",
    "        else:\n",
    "            data.modify_data(\n",
    "                'pc_' + str(args_dct['flip_pc']), flip.flip_windows,\n",
    "                args_dct['flip_window_lst']\n",
    "            )\n",
    "\n",
    "\n",
    "# WRITE RESULTS\n",
    "\n",
    "        # write results to files\n",
    "        data.to_files()\n",
    "\n",
    "\n",
    "# CHROMPLOT\n",
    "\n",
    "if mode  == 'chromplot':\n",
    "    from modules import Plot\n",
    "\n",
    "    plot = Plot('pc_1', stat_var='pc_1_ve', prefix='test', data=data,\n",
    "                chrom=args_dct['start'], start=args_dct['start'], end=args_dct['end'], \n",
    "                color_by=args_dct['color_by'], \n",
    "                metadata_path=args_dct['metadata_path'], \n",
    "                interval=args_dct['interval'], \n",
    "                chromplot_w=config.CHROMPLOT_W,\n",
    "                chromplot_h=config.CHROMPLOT_H,\n",
    "                plot_fmt_lst=args_dct['plot_fmt_lst'],        \n",
    "    )\n",
    "    plot.chromplot()\n",
    "\n",
    "\n",
    "if mode == 'genomeplot':\n",
    "    plot = Plot('pc_1', \n",
    "                run_prefix=args_dct['run_prefix'],\n",
    "                run_id_lst=args_dct['run_id_lst'],\n",
    "                color_by=args_dct['color_by'], \n",
    "                metadata_path=args_dct['metadata_path'], \n",
    "                interval=args_dct['interval'],\n",
    "                genomeplot_w=config.GENOMEPLOT_W,\n",
    "                genomeplot_h=config.GENOMEPLOT_H,\n",
    "                plot_fmt_lst=args_dct['plot_fmt_lst'],\n",
    "    )\n",
    "    plot.genomeplot()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b\n"
     ]
    }
   ],
   "source": [
    "a = []\n",
    "if a == []:\n",
    "    print('b')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
